ul/src/pdu/mod.rs

```diff
@@ -0,0 +1,270 @@
+#[derive(Clone, Eq, PartialEq, PartialOrd, Hash, Debug)]
+pub struct PresentationContextProposed {
+    pub id: u8,
+    pub abstract_syntax: String,
+    pub transfer_syntaxes: Vec<String>,
+}
+
+#[derive(Clone, Eq, PartialEq, PartialOrd, Hash, Debug)]
+pub struct PresentationContextResult {
+    pub id: u8,
+    pub reason: PresentationContextResultReason,
+    pub transfer_syntax: String,
+}
+
+#[derive(Clone, Eq, PartialEq, PartialOrd, Hash, Debug)]
+pub enum PresentationContextResultReason {
+    Acceptance = 0,
+    UserRejection = 1,
+    NoReason = 2,
+    AbstractSyntaxNotSupported = 3,
+    TransferSyntaxesNotSupported = 4,
+}
+
+impl PresentationContextResultReason {
+    fn from(reason: u8) -> Option<PresentationContextResultReason> {
+        let result = match reason {
+            0 => PresentationContextResultReason::Acceptance,
+            1 => PresentationContextResultReason::UserRejection,
+            2 => PresentationContextResultReason::NoReason,
+            3 => PresentationContextResultReason::AbstractSyntaxNotSupported,
+            4 => PresentationContextResultReason::TransferSyntaxesNotSupported,
+            _ => {
+                return None;
+            }
+        };
+
+        Some(result)
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, PartialOrd, Hash, Debug)]
+pub enum AssociationRJResult {
+    Permanent,
+    Transient,
+}
+
+impl AssociationRJResult {
+    fn from(value: u8) -> Option<AssociationRJResult> {
+        match value {
+            0 => Some(AssociationRJResult::Permanent),
+            1 => Some(AssociationRJResult::Transient),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, PartialOrd, Hash, Debug)]
+pub enum AssociationRJSource {
+    ServiceUser(AssociationRJServiceUserReason),
+    ServiceProviderASCE(AssociationRJServiceProviderASCEReason),
+    ServiceProviderPresentation(AssociationRJServiceProviderPresentationReason),
+}
+
+impl AssociationRJSource {
+    fn from(source: u8, reason: u8) -> Option<AssociationRJSource> {
+        let result = match source {
```

Comment:
I'd argue that this 2-level `match` becomes more readable with a single match over the `(source, reason)` pair.

```rust
    let result = match (source, reason) {
        (1, 1) => {
            AssociationRJSource::ServiceUser(AssociationRJServiceUserReason::NoReasonGiven)
        }
        (1, 2) => AssociationRJSource::ServiceUser(
            AssociationRJServiceUserReason::ApplicationContextNameNotSupported,
        ),
        (1, 3) => AssociationRJSource::ServiceUser(
            AssociationRJServiceUserReason::CallingAETitleNotRecognized,
        ),
        (1, 4) | (1, 5) | (1, 6) => {
            AssociationRJSource::ServiceUser(AssociationRJServiceUserReason::Reserved(x))
        }
        (1, 7) => AssociationRJSource::ServiceUser(
            AssociationRJServiceUserReason::CalledAETitleNotRecognized,
        ),
        (1, 8) | (1, 9) | (1, 10) => {
            AssociationRJSource::ServiceUser(AssociationRJServiceUserReason::Reserved(x))
        }
        (1, _) => {
            return None;
        }
        (2, 1) => AssociationRJSource::ServiceProviderASCE(
            AssociationRJServiceProviderASCEReason::NoReasonGiven,
        ),
        (2, 2) => AssociationRJSource::ServiceProviderASCE(
            AssociationRJServiceProviderASCEReason::ProtocolVersionNotSupported,
        ),
        (2, _) => {
            return None;
        }
        (3, 0) => AssociationRJSource::ServiceProviderPresentation(
            AssociationRJServiceProviderPresentationReason::Reserved(0),
        ),
        (3, 1) => AssociationRJSource::ServiceProviderPresentation(
            AssociationRJServiceProviderPresentationReason::TemporaryCongestion,
        ),
        (3, 2) => AssociationRJSource::ServiceProviderPresentation(
            AssociationRJServiceProviderPresentationReason::LocalLimitExceeded,
        ),
        (3, 3) | (3, 4) | (3, 5) | (3, 6) | (3, 7) => {
            AssociationRJSource::ServiceProviderPresentation(
                AssociationRJServiceProviderPresentationReason::Reserved(x),
            )
        }
        _ => {
            return None;
        }
    };
```

Reply:
I agree.